src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py

```diff
@@ -0,0 +1,1028 @@
+            num_precontext = torch.reshape(num_precontext, [-1])
+        else:
+            num_precontext = torch.zeros([num_batch]).int().to(device)
+
+        num_input_contexts = input_ids.shape[1]
+        num_output_contexts = num_input_contexts + num_pasts_contexts
+
+        if past_key_values is not None:
+            if type(past_key_values) is tuple or type(past_key_values) is list:
+                op = []
+                for p in past_key_values:
+                    if type(p) is tuple or type(p) is list:
+                        p = torch.stack(p, dim=1)  # p Shuold be [batch, 2, heads, sequence, kernel]
+                    if p.shape != (
+                        num_batch,
+                        2,
+                        self.config.num_heads,
+                        num_pasts_contexts,
+                        self.config.d_model // self.config.num_heads,
+                    ):  # pasts Should be [batch, layer, 2, heads, sqquence, kernel]
+                        raise ValueError("past_key_values Tuple[Tuple[FloatTensor]]")
+                    op.append(p)
+                pasts = torch.stack(op, dim=1)  # pasts Shuold be [batch, layer, 2, heads, sequence, kernel]
+            if pasts is None or pasts.shape != (
+                num_batch,
+                self.config.num_layers,
+                2,
+                self.config.num_heads,
+                num_pasts_contexts,
+                self.config.d_model // self.config.num_heads,
+            ):  # pasts Should be [batch, layer, 2, heads, sqquence, kernel]
+                raise ValueError("past_key_values Tuple[Tuple[FloatTensor]]")
+        elif self.config.d_spout and spout is not None:
+            pasts = self.spout(spout)
+            pasts = torch.reshape(
+                pasts,
+                [
+                    num_batch,
+                    self.config.num_layers,
+                    2,
+                    self.config.num_heads,
+                    num_pasts_contexts,
+                    self.config.d_model // self.config.num_heads,
+                ],
+            )
+
+        hidden = self.embed_tokens(input_ids)
+
+        if pasts is None:
+            pasts = [None] * self.config.num_layers
+            pos = torch.arange(num_input_contexts).to(device)
+            pos = torch.clip(pos, 0, self.config.num_contexts - 1)
+        else:
+            pasts = [p.squeeze(1) for p in torch.split(pasts, [1] * self.config.num_layers, 1)]
+            pos = torch.arange(num_input_contexts).to(device) + num_pasts_contexts
+            pos = torch.clip(pos, num_pasts_contexts, self.config.num_contexts - 1)
+
+        ppos = (torch.zeros((self.config.d_model, num_input_contexts)).to(device) + pos).transpose(0, 1).long()
+        hidden += torch.gather(self.position_embeddings.weight, dim=0, index=ppos)
+
+        atten_mask = make_attention_mask_torch(num_input_contexts, num_output_contexts, num_precontext)
```

Comment:
let's follow the common variable convention that we use in our codebase : `attention_maks`, `hidden_states`, `positional_embedding` etc